### PR TITLE
Add SIMD implementation of the different statistics

### PR DESCRIPTION
**Pull request by althonos**
## Overview

This PR adds SIMD implementations of the `Identity`, `Gaps`, `Overlaps` and `Similarity` statistics. The way it's organized:

- the `template.h` file contains C++ templates with the implementation of the different functions over an abstract `Vector` type
- the `AVX2.h`, `SSE2.h` and `NEON.h` files contain the concrete implementations for AVX2 (256 bits), SSE2 (128 bits) and NEON (128 bits) vectors
- The choice for a SIMD implementation is done at compile-time in the `Manager` class, which acts as a dispatcher.
 
 I'll add the dynamic dispatch afterwards because it's gonna be a large PR as well, as it needs to pull in and build https://github.com/google/cpu_features to work properly.

## Build

The CMake build will autodetect the best CPU features available. They can be individually disabled to build a version of the binary without SIMD (for testing):

- `cmake . -DDISABLE_SSE2=1 -DDISABLE_AVX2=1` will build without SIMD
- `cmake . -DDISABLE_SSE2=0 -DDISABLE_AVX2=1` will build with SSE2 only
- `cmake . -DDISABLE_SSE2=0 -DDISABLE_AVX2=0` (or just `cmake .`) will build with AVX2 and SSE2 (and effectively use AVX2).

## Tests

To test, I built the SSE2 and the generic binary and compared the outputs for all amino-acid examples with the `-strict`, `-strictplus`, `-gappyout`, `-clusters 5`, `-automated1` and `-resoverlap 0.8 -seqoverlap 75`. Output matched perfectly, but perhaps you'd want to have stricter test cases to compare the expected outputs.

## Benchmarks

I compared the runtime for the AVX2, SSE2 and current implementation on `example.039.AA.bctoNOG.ENOG41099YD.fasta`. The AVX2 code offers a x6 to x12 speedup, with more impressive returns for the overlap statistics. I think it should be feasible to add a SIMD version `Similarity::calculateVectors` for AVX2 using *gather* instructions, but I'm not sure it can be generalized to the other instruction sets.

#### `-strictplus`

<details>
<pre>
Benchmark 1: ./trimal_generic -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -strictplus
  Time (mean ± σ):      1.698 s ±  0.104 s    [User: 1.672 s, System: 0.012 s]
  Range (min … max):    1.494 s …  1.842 s    10 runs
Benchmark 2: ./trimal_sse2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /tmp/test -strictplus
  Time (mean ± σ):     341.3 ms ±  29.0 ms    [User: 332.6 ms, System: 6.2 ms]
  Range (min … max):   306.0 ms … 407.1 ms    10 runs
Benchmark 3: ./trimal_avx2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -strictplus
  Time (mean ± σ):     277.0 ms ±  29.1 ms    [User: 268.3 ms, System: 6.0 ms]
  Range (min … max):   244.5 ms … 328.0 ms    11 runs
Summary
  './trimal_avx2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -strictplus' ran
    1.23 ± 0.17 times faster than './trimal_sse2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /tmp/test -strictplus'
    6.13 ± 0.75 times faster than './trimal_generic -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -strictplus'
</pre>
</details>

#### `-automated1`

<details>
<pre>
Benchmark 1: ./trimal_generic -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -automated1
  Time (mean ± σ):      4.159 s ±  1.177 s    [User: 3.953 s, System: 0.024 s]
  Range (min … max):    3.289 s …  6.908 s    10 runs
Benchmark 2: ./trimal_sse2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /tmp/test -automated1
  Time (mean ± σ):     565.1 ms ±  43.7 ms    [User: 553.9 ms, System: 7.6 ms]
  Range (min … max):   523.5 ms … 672.8 ms    10 runs
Benchmark 3: ./trimal_avx2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -automated1
  Time (mean ± σ):     382.0 ms ±  34.2 ms    [User: 370.9 ms, System: 7.2 ms]
  Range (min … max):   329.2 ms … 424.5 ms    10 runs
Summary
  './trimal_avx2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -automated1' ran
    1.48 ± 0.17 times faster than './trimal_sse2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /tmp/test -automated1'
   10.89 ± 3.23 times faster than './trimal_generic -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -automated1'
</pre>
</details>

#### `-clusters 5`

<details>
<pre>
Benchmark 1: ./trimal_generic -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -clusters 5
  Time (mean ± σ):      1.906 s ±  0.139 s    [User: 1.893 s, System: 0.005 s]
  Range (min … max):    1.662 s …  2.148 s    10 runs
Benchmark 2: ./trimal_sse2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /tmp/test -clusters 5
  Time (mean ± σ):     321.7 ms ±  23.8 ms    [User: 314.8 ms, System: 5.5 ms]
  Range (min … max):   296.2 ms … 364.6 ms    10 runs
Benchmark 3: ./trimal_avx2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -clusters 5
  Time (mean ± σ):     253.5 ms ±  21.0 ms    [User: 245.1 ms, System: 6.3 ms]
  Range (min … max):   220.0 ms … 294.6 ms    10 runs
Summary
  './trimal_avx2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -clusters 5' ran
    1.27 ± 0.14 times faster than './trimal_sse2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /tmp/test -clusters 5'
    7.52 ± 0.83 times faster than './trimal_generic -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -clusters 5'
</pre>
</details>

#### `-resoverlap 0.8 -seqoverlap 75`

<details>
<pre>
Benchmark 1: ./trimal_generic -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -resoverlap 0.8 -seqoverlap 75
  Time (mean ± σ):      3.404 s ±  0.200 s    [User: 3.370 s, System: 0.013 s]
  Range (min … max):    3.124 s …  3.766 s    10 runs
Benchmark 2: ./trimal_sse2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /tmp/test -resoverlap 0.8 -seqoverlap 75
  Time (mean ± σ):     414.4 ms ±  83.4 ms    [User: 396.1 ms, System: 14.7 ms]
  Range (min … max):   354.4 ms … 585.1 ms    10 runs
Benchmark 3: ./trimal_avx2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -resoverlap 0.8 -seqoverlap 75
  Time (mean ± σ):     299.4 ms ±  49.1 ms    [User: 289.2 ms, System: 8.1 ms]
  Range (min … max):   251.2 ms … 394.5 ms    11 runs
Summary
  './trimal_avx2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -resoverlap 0.8 -seqoverlap 75' ran
    1.38 ± 0.36 times faster than './trimal_sse2 -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /tmp/test -resoverlap 0.8 -seqoverlap 75'
   11.37 ± 1.98 times faster than './trimal_generic -in ../dataset/example.039.AA.bctoNOG.ENOG41099YD.fasta -out /dev/null -resoverlap 0.8 -seqoverlap 75'
</pre>
</details>

### CODE REVIEW

File: include/Cleaner.h [hunk 1]
```diff
@@ -468,6 +468,10 @@ class Cleaner {
      */
     Cleaner(Alignment *parent, Cleaner *mold);
 
+    /**
+     \brief Destructor
+     */
+    virtual ~Cleaner() {};
```
Comment (nicodr97) [inline include/Cleaner.h:474]: Why is this necessary? It is not overloaded in any of the SIMD variants

Comment (nicodr97) [inline include/Cleaner.h]: Why is this necessary? It is not overloaded in any of the SIMD variants

File: include/Statistics/Similarity.h [hunk 2]
```diff
@@ -114,7 +114,7 @@ namespace statistics {
             \param sm Similarity matrix pointer to associate.
             \return \b True if sm is valid, \b False if it's null
         */
-        bool setSimilarityMatrix(similarityMatrix * sm);
+        virtual bool setSimilarityMatrix(similarityMatrix * sm);
```
Comment (nicodr97) [inline include/Statistics/Similarity.h]: Why is this necessary? It is not overloaded in any of the SIMD variants

File: include/Statistics/Overlap.h [hunk 3]
```diff
@@ -65,7 +65,7 @@ namespace statistics {
         /**
         \brief Method to calculate overlap between the sequences from the alignment.\n
         */
-        void calculateSeqOverlap();
+        virtual void calculateSeqOverlap();
```
Comment (nicodr97) [inline include/Statistics/Overlap.h:68]: Why is this necessary? It is not overloaded in any of the SIMD variants

File: include/Statistics/Overlap.h
Comment (nicodr97) [inline include/Statistics/Overlap.h:1]: Overlap destructor should be virtual

File: include/Statistics/Identity.h [hunk 4]
```diff
@@ -65,14 +65,14 @@ namespace statistics {
...
-        void calculateSeqIdentity();
+        virtual void calculateSeqIdentity();
 
         /**
         \warning Not In Use
         \brief Method that makes a raw approximation of sequence identity computation.\n
         \note Designed for reducing comparisons for huge alignments.
         */
-        void calculateRelaxedSeqIdentity();
+        virtual void calculateRelaxedSeqIdentity();
```
Comment (nicodr97) [inline include/Statistics/Identity.h]: Why is this necessary? It is not overloaded in any of the SIMD variants

File: include/Statistics/Identity.h
Comment (nicodr97) [inline include/Statistics/Identity.h:1]: Identity destructor should be virtual

File: CMakeLists.txt
Comment (nicodr97) [inline CMakeLists.txt:1]: CMake autodetection does not work. Variables `HAVE_SSE2`, `HAVE_AVX2` and `HAVE_NEON` seems to be undefined when compiling so an error occurs. Moving SIMD includes before checking SIMD compiler support fixes this error

File: scripts/CMake/OBJ-LIB-creator.cmake [hunk 5]
```diff
@@ -60,3 +60,32 @@ add_library(TestsOBJLib             OBJECT ${tests})
...
+  add_compile_definitions(HAVE_AVX2=1)
+  add_library(AVX2OBJLib OBJECT source/Platform/x86/AVX2.cpp)
+  if(NOT AVX2_C_FLAGS STREQUAL " ")
+    target_compile_options(AVX2OBJLib PRIVATE "${AVX2_C_FLAGS}")
+  endif()
+endif()
+
+# NEON
+if (HAVE_NEON)
+  message(STATUS "Detected compiler support for AVX2 CPU extensions.")
```
Comment (nicodr97) [inline scripts/CMake/OBJ-LIB-creator.cmake]: Replace `AVX2 CPU` with `NEON CPU`

**Review by nicodr97 — CHANGES_REQUESTED** [review]
Besides this comments, I am testing the program in its different versions with a dataset of MSAs (not the small ones in this repo) just to be sure these implementations are not affecting the final result.

File: include/Statistics/Similarity.h (hunk 2 above)
Comment (althonos) [reply include/Statistics/Similarity.h]: That's a remain from the Python port, I'll remove it.

File: include/Statistics/Overlap.h (hunk 3 above)
Comment (althonos) [reply include/Statistics/Overlap.h:68]: I thought about adding a SIMD overload as well, but since it's not used anywhere I left it as-is, I'll remove the `virtual`.

File: CMakeLists.txt
Comment (althonos) [reply CMakeLists.txt:1]: It's working on my end, so I guess this may be a CMake version thing? In any case I moved the includes earlier in the CMake list.

File: include/Statistics/Identity.h (hunk 4 above)
Comment (althonos) [reply include/Statistics/Identity.h]: Same as `calculateSeqOverlap`, it's not used anywhere so I ended up not adding a SIMD variant, but if you feel it would be useful I can always add one later.

File: include/Platform/template.h [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+      for (; k < alig->originalNumberOfResidues; k++) {
+        int nongapi = (datai[k] != indet) && (datai[k] != '-');
+        int nongapj = (dataj[k] != indet) && (dataj[k] != '-');
+        hits_u8[k] += ((nongapi && nongapj) || (datai[k] == dataj[k]));
+      }
+
+      // we can process up to UCHAR_MAX sequences, otherwise hits_u8[k]
+      // may overflow, so every UCHAR_MAX iterations we transfer the
+      // partial hit counts from `hits_u8` to `hits`
+      if ((j % UCHAR_MAX) == 0) {
```
Comment (nicodr97) [inline include/Platform/template.h]: If the sequence to be compared (`i`) is a multiple of `UCHAR_MAX` `hits_u8` does not get reset. I add comments with the code to consider this and the rest of cases.

File: include/Platform/template.h [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+  // for each sequence in the alignment, computes its overlap
+  for (int i = 0; i < alig->originalNumberOfSequences; i++) {
+
+    // reset hits count
+    memset(&hits[0], 0, alig->originalNumberOfResidues * sizeof(uint32_t));
+    memset(&hits_u8[0], 0, alig->originalNumberOfResidues * sizeof(uint8_t));
+
+    const uint8_t *datai =
+        reinterpret_cast<const uint8_t *>(alig->sequences[i].data());
+
```
Comment (nicodr97) [inline include/Platform/template.h:347]: `int processedSequences = 0;`

File: include/Platform/template.h (hunk 6 above)
Comment (nicodr97) [inline include/Platform/template.h]: `if ((processedSequences % UCHAR_MAX) == 0) {`

File: include/Platform/template.h [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+      }
+
+      // we can process up to UCHAR_MAX sequences, otherwise hits_u8[k]
+      // may overflow, so every UCHAR_MAX iterations we transfer the
+      // partial hit counts from `hits_u8` to `hits`
+      if ((j % UCHAR_MAX) == 0) {
+        for (k = 0; k < alig->originalNumberOfResidues; k++)
+          hits[k] += hits_u8[k];
+        memset(hits_u8, 0, alig->originalNumberOfResidues * sizeof(uint8_t));
+      }
```
Comment (nicodr97) [inline include/Platform/template.h:396]:
```
}

processedSequences++;
```

**Review by nicodr97 — CHANGES_REQUESTED** [review]
Once this is fixed the PR can be merged

**Comment by althonos** [comment]
Good catch, thank you!

**Review by nicodr97 — APPROVED** [review]

**Comment by althonos** [comment]
I'll come back to you when I have more time to implement the dynamic dispatching :+1:

**Comment by althonos** [comment]
Btw I made a small mistake in the `cmake` instructions, you may want to update the `README.md`! I updated the instructions in the post above.

**Comment by althonos** [comment]
@nicodr97 : i have a little bit of time to implement the dynamic dispatch now, on which branch should I start writing it? :)

**Comment by nicodr97** [comment]
Sorry for not answering before, I was not aware of this message.
Actually, I am trying to implement SIMD instructions (starting with AVX2) in ```Identity::calculateSeqIdentity``` and ```simd::calculateSimilarityVectors```. I didn't do it yet, I was just understanding how you did it in the other parts of the code. What I already did is to use OpenMP, but I would like to do both implemenations together, since I saw there is an simd directive in OpenMP.
Anyway, you can work on branch 2.0_RC for the dynamic dispatch and we can discuss the parallelizations when you are available.
